### ISSUE TITLE
Example of convolutional maxout layer

### ISSUE THREAD

**Issue by GregAtHeron** [benanne/kaggle-galaxies#2]
I'm playing around with your code to get a better insight into the inner-workings on convnets, and was curious as to how maxout on convolutional layers was implemented for you, even if unused in your final scripts. 

(interestingly I've found that turning dropout on bottom convolution layers tends to make the problem over-fit wildly early on, which is interesting, if baffling)

**Comment by benanne** [comment]
I just used a regular convolutional layer followed by a layers.FeatureMaxPoolingLayer, which is also how the maxout in the dense layers is implemented. It takes a 'feature_dim' argument which indicates the axis along which the pooling will take place, so you just have to set that to the feature maps axis (I think feature_dim=1 should work for both dense and conv layers in my particular implementation).

Regarding your observation about dropout in the bottom convolutional layers, I've seen something similar happen recently, and I have no explanation for it either. Baffling indeed.

**Comment by GregAtHeron** [comment]
Thank you for the response.

I'm probably misunderstanding what maxout is doing then in convolutional layers, and I found the original paper's explanation of maxout on convnets quite short.

In a normal convolutional layer with reLU, each patch of an image (for example RGB_N_N for a square color image) is put through the convolutional kernel, resulting in what essentially is a dot product of two 3D tensors, if we're ignoring minibatch dimension(height \* width \* (k feature maps from previous layer)). The resulting dot product is then put through reLU, resulting in that particular feature map pixel value.

I thought that MaxOut instead takes the max over the k featuremaps dot producted with the k kernels (essentially ignoring the input from the other k-1 kernel maps and just returning that value sans reLU/tanh).

This interpretation doesn't seem to square with being able to run the regular convolution layer over the entire patch(with all k feature maps from the previous layer) and then max over channels. Or I'm just not yet understanding the code.

Sorry for the questions and the lack of clarity; I have yet to find a great explicit resource for how things should be computed, and I'm learning by myself in industry.

**Comment by benanne** [comment]
I'm not sure I fully understand your explanation, but a maxout unit is just taking the maximum over two or more linear filter correlations, just like a ReLU takes the maximum over a linear filter correlation and zero. So you can effectively implement it by pooling a bunch of linear filters together. It doesn't matter whether those filters are dense or convolutional (or some other connectivity pattern).

**Comment by GregAtHeron** [comment]
As I thought, I must be having a fundamental misunderstanding. I'll start with the reLU case. 

"a ReLU takes the maximum over a linear filter correlation and zero"
Let's suppose we have k feature maps coming out of the previous layer. Our weight matrix corresponding to a single filter on this new layer is used to do the linear filter correlation on k_d_d pixels, where d = convolutional filter height and width. The resulting value is a scalar, which is then put through max(scalar, 0), resulting in a single pixel value in a new feature map.

Is this correct?

**Comment by benanne** [comment]
No, the resulting value from a convolution with a d_d filter is not going to be a single scalar value, unless the input size also happens to be d_d.

If your input has shape w*h, then the output of the (valid) convolution is going to have shape (w - d + 1, h - d + 1) and the max(., 0) function is then applied to this matrix in an elementwise fashion.

**Comment by GregAtHeron** [comment]
I believe we are on the same page(I was describing it in an "elementwise" fashion). 

Just to be pedantic and simplistic, an RGB image(3_w_h) will be convolved into a 2D matrix (1_(w-d+1)_(h-d+1)) with a valid convolution(k_d_d filter? or d*d done k times?). The resulting matrix will be elementwise non-negative as per reLU.

From there how is this converted into MaxOut?

Thanks again! There are so many terms in convnet land, makes it really hard to communicate unless you know the lingo.

**Comment by benanne** [comment]
This is getting a bit confusing, let's define more clearly what we both mean with each of these numbers/letters.

Let's take the image case (i.e. 2D convolutions). The input data will look something like this:
(minibatch_size, num_input_channels, image_width, image_height)

The filters will look like this:
(num_filters, num_input_channels, filter_width, filter_height)

So the output shape will be:
(minibatch_size, num_filters, image_width - filter_width + 1, image_height - filter_height + 1)

Now we can either apply f(x) = max(x, 0) elementwise to this 4D output tensor to implement rectified linear units. Alternatively, we can apply max-pooling along the 'num_filters' axis. For example, if we decide to pool pairs of filters together, the output will end up being:
(minibatch_size, num_filters / 2, image_width - filter_width + 1, image_height - filter_height + 1)

I hope that clears things up a bit. Maxout is nothing more than max-pooling along the 'filters' dimension, instead of along the topological dimensions of the input (i.e. width and height). That's really all there is to it. And it doesn't matter whether you apply this to convolutional feature maps, or to the output of a fully connected layer. The connectivity between the layers is completely orthogonal to the nonlinearity that is used.

**Comment by GregAtHeron** [comment]
Alright, I think that's pretty clear, but I'll take another stab at this, in layman's terms:

If the Convolutional Layer was set to output 128 feature maps(num_filters), the MaxOut is just pooling pairs(sets) of these together with the max() function. So pooling pairs would result in 64 feature maps for the next layer! Now that I'm typing it out, I see the similarity to vanilla max pooling!

Sound right?

edit: I had MaxOut complete backwards, now it makes sense in your code! 

Much thanks!

**Comment by benanne** [comment]
Yep, that's exactly it. No problem!

**Comment by GregAtHeron** [comment]
_lightbulb just went off_

I was wrong about the fully connected layers too. Same issue(I was maxing over inputs, not activations). No wonder your statements just further confused me. Just got it. Awesome!
